docs/labs/format-strings.html

```diff
@@ -0,0 +1,136 @@
+  examples:
+  - - "'some format'.format(user_input=user_input, event=event)"
+  - - "'some format'.format(user_input, event)"
+- absent: \{(0|format_string)\}
+  text: Make sure the braces for the format_string replacement field are in the format string
+  examples:
+  - - "'{event}'.format(format_string=user_input)"
+  - - "'{0}.format(user_input)"
+- absent: \{(1|event)\}
+  text: Make sure the braces for the event replacement field are in the format string
+  examples:
+  - - "'{format_string}, {event.level}'.format(format_string=user_input, event=event)"
+  - - "'{0}, {1.level}'.format(user_input, event)"
+- absent: \{(1\.level|event\.level)\}
+  text: Make sure the level attribute of the event replacement field is in the format string
+  examples:
+  - - "'{format_string}, {event.level}'.format(format_string=user_input, event=event)"
+  - - "'{0}, {1.level}'.format(user_input, event)"
+# debug: true
+</script>
+</head>
+<body>
+<!-- For GitHub Pages formatting: -->
+<div class="container-lg px-3 my-5 markdown-body">
+<h1>Lab Exercise Format Strings and Templates</h1>
+<p>
+This is a lab exercise on developing secure software.
+For more information, see the <a href="introduction.html" target="_blank">introduction to
+the labs</a>.
+
+<p>
+<h2>Task</h2>
+<p>
+<b>Practice using string templates in a secure way.</b>
+
+<p>
+<h2>Background</h2>
+<p>
+In this exercise, we'll adjust our string formatting so that it doesn't allow a user to control
+the <a href="https://docs.python.org/3/tutorial/inputoutput.html#the-string-format-method"><tt>
+format string</tt></a>. If a user can control the <tt>format string</tt> they can access
+variables which they shouldn't. Particularly if those variable's values can be returned to the user
+as output, it could lead to information disclosure beyond what was intended by the developer.
+
+<p>
+<h2>Task Information</h2>
+<p>
+
+<p>
+Please change the code below so the string formatting cannot disclose arbitrary
+program values. The server-side program is written in Python and allows a user to specify a
+<tt>format string</tt> to control the output format of an event.
+
+<p>
+You could adjust the program so that it only formats the event, and does not include any user input,
+However it is considered safe to include user input in the output as long as they cannot control
+the <tt>format string</tt> itself.
+
+<p>
+Adjust the value returned by the <tt>format_event</tt> function so the the user controlled
+<tt>user_input</tt> variable is only used as a
+<a href="https://docs.python.org/3/library/string.html#format-string-syntax"><tt>replacement field</tt></a>
+and is not used as the <tt>format string</tt>.
+
+<p>
+Use the “hint” and “give up” buttons if necessary.
+
+<p>
+<h2>Interactive Lab (<span id="grade"></span>)</h2>
+<p>
+<form id="lab">
+<pre><code
+> # Application configuration which should be kept secret from a user
+CONFIG = {
+    'SECRET_KEY': 'super secret key'
```

Comment:
Don't do that! That would be a secret key embedded in source code, which is *itself* a vulnerability.